8325520: Vector loads and stores with indices and masks incorrectly compiled

dafedafe opened the pull request:
# Issue
When loading multiple vectors using indices or masks (e.g. `LongVector::fromArray(LongVector.SPECIES_256, storage, 0, indices, 0` or `LongVector::fromArray(LongVector.SPECIES_256, storage, 0, longMask)`) there is an error in the C2 compiled code that makes different vectors be treated as equal even though they are not.

# Causes
On vector-capable platforms, vector loads with masks and indices (for Long, Integer, Float and Double) create specific nodes in the ideal graph (i.e. `LoadVectorGather`, `LoadVectorMasked`, `LoadVectorGatherMasked`). Vector loads without mask or indices are mapped as `LoadVector` nodes instead.
The same is true for `StoreVector`s.
When running GVN loops we can get to the situation where we check if a Load node is preceded by a Store of the same address to be able to replace the Load with the input of the Store (in `LoadNode::Identity`). Here we call

https://github.com/openjdk/jdk/blob/87e864bf21d71daae4e001ec4edbb4ef1f60c36d/src/hotspot/share/opto/memnode.cpp#L1258

where we do an extra check for types if we deal with vectors but we don’t make sure that neither masks nor indices interfere.
Similarly, in `StoreNode::Identity`  we first check if there is a Load and then a Store:

https://github.com/openjdk/jdk/blob/87e864bf21d71daae4e001ec4edbb4ef1f60c36d/src/hotspot/share/opto/memnode.cpp#L3509-L3515

but we don’t make sure that there are no masks or indices. 
A few lines below, we check if there are 2 stores for the same value in a row. We need to check for masks and indices here too but in this case we can include these cases if the masks and indices of the vector stores are equivalent.

# Solution
To avoid folding `Load`- and `StoreVector`s with masks and indices we add a specific `store_Opcode` method to `LoadVectorGatherNode`, `LoadVectorMaskedNode` and `LoadVectorGatherMaskedNode` that doesn’t return a store opcode but instead returns -1. In this way, the checks in `MemNode::can_see_stored_value`

https://github.com/openjdk/jdk/blob/87e864bf21d71daae4e001ec4edbb4ef1f60c36d/src/hotspot/share/opto/memnode.cpp#L1164-L1166

and `StoreNode::Identity`

https://github.com/openjdk/jdk/blob/87e864bf21d71daae4e001ec4edbb4ef1f60c36d/src/hotspot/share/opto/memnode.cpp#L3509-L3515

will fail if masks or indices are used.
For 2 stores of the same value we instead check for mask and indices equality.

Regression tests for all versions of `Load/StoreVectorGather/Masked` have been added too.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325520](https://bugs.openjdk.org/browse/JDK-8325520): Vector loads and stores with indices and masks incorrectly compiled (**Bug** - P3)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18347/head:pull/18347` \
`$ git checkout pull/18347`

Update a local copy of the PR: \
`$ git checkout pull/18347` \
`$ git pull https://git.openjdk.org/jdk.git pull/18347/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18347`

View PR using the GUI difftool: \
`$ git pr show -t 18347`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18347.diff">https://git.openjdk.org/jdk/pull/18347.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18347#issuecomment-2055838191)

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -1127,6 +1143,41 @@ Node* StoreVectorMaskedNode::Ideal(PhaseGVN* phase, bool can_reshape) {
 }
+Node* StoreVectorMaskedNode::Identity(PhaseGVN* phase) {
+  Node* value = StoreVectorNode::Identity(phase);
+  if ((value != this) &&  (value->is_StoreVectorMasked()) && (in(MemNode::ValueIn + 1)->eqv_uncast(value->in(MemNode::ValueIn + 1)))) {
```
`MemNode::ValueIn + 1` is there no direct value for it? If not: maybe create one?

dafedafe replied:
Thanks for the comments @eme64. 
I've created one more constant for the "+ 2" case (`OffsetsMask`) and reused the one there for the "+ 1" case (`OopStore`).

eme64 reviewed (commented):
Can you please fix the whitespace issues, it is a bit difficult to read now on GitHub ;)

eme64 commented:
I see that `LoadVectorNode` has no additional slots, but then `LoadVectorMaskedNode` and `LoadVectorGatherMaskedNode` simply do `add_req`. Why not just compare these extra slots (as many as there are) in `LoadVectorNode::Identity`, as well as checking that their `Opcode` is the same? Would that not be simpler and accomplish the same?

eme64 commented on test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:
```diff
@@ -0,0 +1,183 @@
+ * @summary Don't allow folding of load/store of vectors when using incompatible indices or mask
+ * @modules jdk.incubator.vector
+ *
+ * @run main/othervm -XX:UseAVX=3 compiler.vectorapi.VectorLoadGatherFoldingTest
```
I would also make sure that you don't require platform specific things here. This could also run on any other platform that supports the Vector API, right? And you will need at least one run without any flags.

dafedafe replied:
You're right. I've relaxed the require (the max vector size shouldn't be relevant) and added `sve`. I'm not sure if there is a way to make it more generic (all platforms that support vectors?).

dafedafe replied:
True. It might test the Vector API so often but in the end we need to test the scalar too. Fixed.

dafedafe commented:
> Can you please fix the whitespace issues, it is a bit difficult to read now on GitHub ;)

Sorry, I let them slip in. Fixed.

dafedafe commented on test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:
```diff
@@ -0,0 +1,183 @@
+ * @run main/othervm -XX:UseAVX=3 compiler.vectorapi.VectorLoadGatherFoldingTest
+ */
+
+public class VectorLoadGatherFoldingTest {
```
Yep, there are actually `Store` tests already but I forgot to adapt the name. Fixed now.
I've tried to use the IR framework to check for folded/non folded nodes but couldn't make it reliable enough (in the end it didn't test more than the current test). So I decided to go back to the actual "regression" test which reproduces the original issue.

eme64 commented on test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:
```diff
@@ -0,0 +1,183 @@
+ * @summary Don't allow folding of load/store of vectors when using incompatible indices or mask
+ * @modules jdk.incubator.vector
+ *
+ * @run main/othervm compiler.vectorapi.VectorGatherMaskFoldingTest
```
```suggestion
 * @run main compiler.vectorapi.VectorGatherMaskFoldingTest
```
I think you only need the `othervm` if you also use flags in the `@run` statement.

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -1127,6 +1143,41 @@ Node* StoreVectorMaskedNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+
+Node* StoreVectorScatterMaskedNode::Identity(PhaseGVN* phase) {
+  Node* value = StoreVectorNode::Identity(phase);
+  if ((value != this) &&  (value->is_StoreVectorScatterMasked()) &&
+    (in(MemNode::OopStore)->eqv_uncast(value->in(MemNode::OopStore))) &&
+    (in(MemNode::OffsetsMask)->eqv_uncast(value->in(MemNode::OffsetsMask)))) {
```
```suggestion
  if ((value != this) &&  (value->is_StoreVectorScatterMasked()) &&
      (in(MemNode::OopStore)->eqv_uncast(value->in(MemNode::OopStore))) &&
      (in(MemNode::OffsetsMask)->eqv_uncast(value->in(MemNode::OffsetsMask)))) {
```

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -1127,6 +1143,41 @@ Node* StoreVectorMaskedNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+
+Node* LoadVectorGatherMaskedNode::Identity(PhaseGVN* phase) {
+  Node* value = LoadVectorNode::Identity(phase);
+  if ((value != this) &&  (value->is_LoadVectorGatherMasked()) &&
+    (in(MemNode::ValueIn)->eqv_uncast(value->in(MemNode::ValueIn))) &&
+    (in(MemNode::OopStore)->eqv_uncast(value->in(MemNode::OopStore)))) {
```
```suggestion
  if ((value != this) &&  (value->is_LoadVectorGatherMasked()) &&
      (in(MemNode::ValueIn)->eqv_uncast(value->in(MemNode::ValueIn))) &&
      (in(MemNode::OopStore)->eqv_uncast(value->in(MemNode::OopStore)))) {
```

eme64 reviewed (changes requested):
Issues with indentation

eme64 commented:
> It seems that your code would now disallow such a case, because you always check that the Ideal node you get back is of the same type as the this node. Am I right about that? Is that intended?

If I am right that you would have made such a optimization impossible, that probably means that our tests don't have an IR test that cover this case. You would definitely need to add such IR tests, otherwise we don't know if we are getting regressions. You will probably also have to run this patch through performance testing eventually.

> I've tried to use the IR framework to check for folded/non folded nodes but couldn't make it reliable enough (in the end it didn't test more than the current test). So I decided to go back to the actual "regression" test which reproduces the original issue.

Do you know what is the issue with reliability for the IR rules? Why did it not always work?

dafedafe commented:
> I think you checks should also not be done **after** we already create a new node, but ideally before we create the new node. That is the normal pattern I see everywhere. So then you would probably have to dig into `MemNode::can_see_stored_value` and other places, to see where you would need to do your additional checks. There are already some checks for vector-type there, so that would be one intuitive starting-point.

I've removed the `Identity` overrides and changed the checks in `MemNode::can_see_stored_value` as you suggested (`Identity` methods were also a bit too restrictive). I also had to add checks to `StoreNode::Identity` here

https://github.com/openjdk/jdk/blob/7eb78e332080df3890b66ca04338a4ba69af45eb/src/hotspot/share/opto/memnode.cpp#L2799-L2814

dafedafe commented:
> If I am right that you would have made such a optimization impossible, that probably means that our tests don't have an IR test that cover this case. You would definitely need to add such IR tests, otherwise we don't know if we are getting regressions. You will probably also have to run this patch through performance testing eventually.

I've transformed the tests to add IR tests as well. The issue with them seems to be related with [JDK-8302459](https://bugs.openjdk.org/browse/JDK-8302459) (basically there are some missing cleanups when performing late inlining). So, for now the tests force a cleanup at every step (`-XX:+IncrementalInlineForceCleanup`).

eme64 commented:
@dafedafe Nice, I think this already looks much better. Let me know if/when you want me to look at it again ;)

dafedafe commented:
> @dafedafe Nice, I think this already looks much better. Let me know if/when you want me to look at it again ;)

Thanks a lot @eme64!

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -2801,7 +2827,35 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
+        // Regular store (no offsets or mask)
+        } else {
+          result = mem;
+        }
```
You already have the condition `val->as_Load()->store_Opcode() == Opcode()`.
So once you have `is_StoreVectorScatter()`, I think `val->is_LoadVectorGather()` is implied, no?

Oh wow. I think actually that this is another bug here: we only have
`virtual int store_Opcode() const { return Op_StoreVector; }` for `LoadVectorNode`, but not for all masked/gather/scatter vector nodes! I think that should be fixed.

That would also simplify your code.

dafedafe replied:
I’ve updated the `store_Opcode` methods. Actually I’ve used them to always avoid being the same as a Store opcode (I’ve used the Load own opcodes (a bit hacky) but that can be changed). See the comment below.

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -2801,7 +2827,35 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
+        // Handle StoreVector with offsets and masks
+        // Ensure offsets match
+        if (is_StoreVectorScatter()) {
+          const Node* offsets = as_StoreVectorScatter()->in(StoreVectorScatterNode::Offsets);
+          if (val->is_LoadVectorGather() && offsets->eqv_uncast(val->as_LoadVectorGather()->in(LoadVectorGatherNode::Offsets))) {
```
```suggestion
          const Node* offsets_store = as_StoreVectorScatter()->in(StoreVectorScatterNode::Offsets);
          const Node* offsets_load = val->as_LoadVectorGather()->in(LoadVectorGatherNode::Offsets);
          if (offsets_store->eqv_uncast(offsets_load)) {
```
As explained below, `val->as_Load()->store_Opcode() == Opcode()` (with the fix I described) would imply that `val->is_LoadVectorGather()`.

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -1163,13 +1163,39 @@ Node* MemNode::can_see_stored_value(Node* st, PhaseValues* phase) const {
         const TypeVect* out_vt = as_LoadVector()->vect_type();
         if (in_vt != out_vt) {
           return nullptr;
         }
```
I see there is a vector type check here. Do we not need that for the code in `StoreNode::Identity`? "Normal" stores like `StoreB` and `StoreI` have the type implicit, but for vector nodes, this type is hidden in the `vect_type()`, so I suspect you need to check it.

I imagine a scenario where we store a float-vector, and then read from the same address as int-vector. Is that ok, or would we need a ReinterpretCast node? I'm not sure, but it would be worth trying to create some tests to check that.

eme64 replied:
Can you actually do that: store a float-vector to an int-array? Or is that maybe only possible with Unsafe somehow? Or maybe completely impossible?

dafedafe replied:
The API doesn't seem to allow this. I'm not sure of what one could do with Unsafe. Would we have to take potential Unsafe operations into account?

eme64 reviewed (changes requested):
@dafedafe Nice work! I left a few comments ;)
I wonder if there cannot be some shared method now, since the code in both cases looks quite similar.

TobiHartmann commented on test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:
```diff
@@ -0,0 +1,897 @@
+ * @summary Don't allow folding of Load/Store vectors when using incompatible indices or masks
+ * @modules jdk.incubator.vector
+ *
+ * @run main compiler.vectorapi.VectorGatherMaskFoldingTest
```
```suggestion
 * @run driver compiler.vectorapi.VectorGatherMaskFoldingTest
```

TobiHartmann commented on test/hotspot/jtreg/compiler/vectorapi/VectorGatherMaskFoldingTest.java:
```diff
@@ -0,0 +1,897 @@
+    // LongVector tests
+
+    @Test
+    @Warmup(10000)
```
Since all tests use the same warmup, I would suggest to set it once via `testFrameworkobject.setDefaultWarmup(10000)`, see https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/lib/ir_framework/README.md

dafedafe replied:
👍 thanks for the suggestion!

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -2801,7 +2827,35 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
       val->in(MemNode::Memory )->eqv_uncast(mem) &&
       val->as_Load()->store_Opcode() == Opcode()) {
-    result = mem;
+        // Handle StoreVector with offsets and masks
```
Also: the indendation is not right: it should only be 2 spaces from the `if`.

dafedafe replied:
Yep, fixed, thanks.

eme64 commented:
I've randomly thought about this change again (while trying to sleep yesterday...). And started worrying about this kind of case:

```
import jdk.incubator.vector.*;

public class Test {
    private static final VectorSpecies<Integer> I_SPECIES = IntVector.SPECIES_MAX;

    public static void main(String[] args) {
        // create mask
        boolean[] intMask = new boolean[I_SPECIES.length()];
        for (int i = 0; i < intMask.length; i++) {
            intMask[i] = (i % 2 == 0);
        }
        int[] intArray1 = new int[I_SPECIES.length()];
        int[] intArray2 = new int[I_SPECIES.length()];
        int[] intArray3 = new int[I_SPECIES.length()];
        for (int i = 0; i < intArray1.length; i++) {
            intArray1[i] = i;
            intArray2[i] = -2 * i;
            intArray3[i] = 0;
        }

        for (int i = 0; i < 10_000; i++) {
            test1(intMask, intArray1, intArray2, intArray3);
        }

        for (int i = 0; i < intArray1.length; i++) {
            System.out.println("i: " + i + " " + intArray1[i] + " " +intArray2[i] + " " + intArray3[i]);
        }
    }

    static void test1(boolean[] intMask, int[] intArray1, int[] intArray2, int[] intArray3) {
        VectorMask<Integer> intVectorMask = VectorMask.fromArray(I_SPECIES, intMask, 0);

        // Load values:    0   1   2   3   4   5 ...
        IntVector a = IntVector.fromArray(I_SPECIES, intArray1, 0);

        // Store, but only every second value:
	// Store:          0   x   2   x   4   x ...
	// Already there:  0  -2  -4  -6  -8 -10 ...
	// Result:         0  -2   2  -6   4 -10 ...
        a.intoArray(intArray2, 0, intVectorMask);

        // Load, but we cannot just take the value from vector a, since we mask it, and where the mask is
	// off, we must have zero.
	// Load:           0   0   2   0   4   0 ...
        IntVector b = IntVector.fromArray(I_SPECIES, intArray2, 0, intVectorMask);
        b.intoArray(intArray3, 0);
    }
}
```

This is not a bug yet, I think, but it could be if you do the change I suggested with `store_Opcode()`.
Currently, if we have:
```
 2035  StoreVectorMasked  === 934 7 2031 1979 1956  |893  [[ 2034 2041 2066 ]]  @int[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact[0] *, idx=7; mismatched  Memory: @int[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact[0] *, idx=7; !jvms: IntVector::intoArray0Template @ bci:50 (line 3556) IntMaxVector::intoArray0 @ bci:9 (line 911) IntVector::intoArray @ bci:53 (line 3269) Test::test1 @ bci:26 (line 34)
 2041  LoadVectorMasked  === 934 2035 2031 1956  |893  [[ 1767 2066 1750 ]]  @int[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact[0] *, idx=7; mismatched #vectorz[16]:{int} !jvms: IntVector::fromArray0Template @ bci:53 (line 3455) IntMaxVector::fromArray0 @ bci:11 (line 874) IntVector::fromArray @ bci:32 (line 2986) Test::test1 @ bci:36 (line 35)
```
The load does not succeed in  `MemNode::can_see_stored_value`, because the load's `store_Opcode() == Op_StoreVector`, and not `Op_StoreVectorMasked`. But if you were to implement `LoadVectorMasked::store_Opcode() const { return Op_StoreVectorMasked; }`, then you have to be careful:
A masked load does not necessarily return the same as the masked store's input value. That input value is not yet masked, but the loaded value needs to be masked.

But it seems to me that you can actually never have a successful `MemNode::can_see_stored_value` case for masked operations, with your current code. It would always fail the `store_Opcode() == st->Opcode()` check. And for that gives the correct result, but it is still a bit strange that we don't override the `store_Opcode` for the masked/offset vector stores.

I don't know which way you want to go now. I these options:
- Keep disallowing masked load/store "look-throughs".
  - Do that by having the "incorrect" `store_Opcode` as now. The downside is that the "offset only" case does not manage to do the look-through, even though that would be correct.
  - OR: have the correct `store_Opcode`, which allows the look-through for the "offset only" case. But then explicitly check for the masked cases, and disallow those.
- Implement a special look-through, where you apply the mask with some blend/select/masked operation on the store input value, which simulates the masked load (i.e. you need to put zeros where the mask is off).

Not sure if this is all very clear, feel free to ask.

dafedafe commented:
> The load does not succeed in `MemNode::can_see_stored_value`, because the load's `store_Opcode() == Op_StoreVector`, and not `Op_StoreVectorMasked`. But if you were to implement `LoadVectorMasked::store_Opcode() const { return Op_StoreVectorMasked; }`, then you have to be careful: A masked load does not necessarily return the same as the masked store's input value. That input value is not yet masked, but the loaded value needs to be masked.
> 
> But it seems to me that you can actually never have a successful `MemNode::can_see_stored_value` case for masked operations, with your current code. It would always fail the `store_Opcode() == st->Opcode()` check. And for that gives the correct result, but it is still a bit strange that we don't override the `store_Opcode` for the masked/offset vector stores.
> 
> I don't know which way you want to go now. I these options:
> 
> * Keep disallowing masked load/store "look-throughs".
>   
>   * Do that by having the "incorrect" `store_Opcode` as now. The downside is that the "offset only" case does not manage to do the look-through, even though that would be correct.
>   * OR: have the correct `store_Opcode`, which allows the look-through for the "offset only" case. But then explicitly check for the masked cases, and disallow those.
> * Implement a special look-through, where you apply the mask with some blend/select/masked operation on the store input value, which simulates the masked load (i.e. you need to put zeros where the mask is off).
> 
> Not sure if this is all very clear, feel free to ask.

Yep it is actually very clear, thanks a lot @eme64! Actually your example is with masks, but there is a similar problem with offsets as well. Think of what happens if the offset array is shorter than the length of the array or if there are duplicate indices in the offset array: the result of a load-store or a store-load sequence doesn’t produce the same original array or vector.
So actually to avoid folding in those 2 cases (load->store and store->load) for now I made specific `store_Opcode` implementations that are never the same as a Store opcode, so that the `store_Opcode() == st->Opcode()` and `val->as_Load()->store_Opcode() == Opcode()` is always false (I’ve used the opcode of the `LoadNode` itself but possibly there is a better solution).
As you hint one could also think of a solution involving substituting load/store masked vector nodes with “simple” vector mask nodes and load/store gather/scatter vector nodes with “simple” vector shuffle nodes (if these exist) but I’m not sure how much this would bring and anyway for this issue I would keep the fix simple. What do  you think?

eme64 commented:
Nice, ah you are right, there can be issues with mask-only cases as well!

It would be great if you had tests that exactly exercise these "bad" examples, where it looks like we might optimize, but it would be wrong.

I'll look at your `store_Opcode` changes now...

dafedafe commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -910,6 +911,10 @@ class LoadVectorGatherNode : public LoadVectorNode {
   }
+  virtual int store_Opcode() const {
+    // Ensure it is different from any store opcode
+    return Op_LoadVectorGather;
```
OK. I was just a bit concerned that a check for equality between 2 `store_Opcode` could be true because they are both -1 but this shouldn’t happen. Changed.

dafedafe commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -3511,7 +3512,35 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
+      const StoreVectorScatterMaskedNode* stv = as_StoreVectorScatterMasked();
+      const Node* offsets = stv->in(StoreVectorScatterMaskedNode::Offsets);
+      const Node* mask = stv->in(StoreVectorScatterMaskedNode::Mask);
+      if (mem->is_StoreVectorScatterMasked()) {
```
Yep, I removed this useless `if` and 2 other terms in the ifs before that. I'm just not sure of what you mean with
> if you extracted the offsets and masks in parallel.

dafedafe commented:
> It would be great if you had tests that exactly exercise these "bad" examples, where it looks like we might optimize, but it would be wrong.

Yep, good idea. I've added a few tests to check for those cases (load-store with duplicate offsets and  store-load with masks). Thanks @eme64!

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -1164,8 +1164,9 @@ Node* MemNode::can_see_stored_value(Node* st, PhaseValues* phase) const {
-      // LoadVector/StoreVector needs additional check to ensure the types match.
+      // LoadVector/StoreVector need additional checks
       if (st->is_StoreVector()) {
+        // Ensure that types match
```
To reduce  noise, you could revert these comment changes, up to you.

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -3511,7 +3515,43 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
       mem->in(MemNode::ValueIn)->eqv_uncast(val) &&
       mem->Opcode() == Opcode()) {
-    result = mem;
+    // Not a vector
```
```suggestion
```
Redundant comment, the next line literally says as much ;)

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -3511,7 +3515,43 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
+          }
+        // Regular store (no offsets or mask)
+        } else {
+          result = mem;
```
```suggestion
          assert(Opcode() = Op_StoreVector, "just a plain vector store, no offset or mask");
          result = mem;
```
Turning comments into asserts is preferable I would say.

dafedafe replied:
Good idea! Thanks!

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -3511,7 +3515,43 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
+          const Node* mask = stv->in(StoreVectorScatterMaskedNode::Mask);
+          const StoreVectorScatterMaskedNode* svgm = mem->as_StoreVectorScatterMasked();
+          if (offsets->eqv_uncast(svgm->in(StoreVectorScatterMaskedNode::Offsets)) &&
+            mask->eqv_uncast(svgm->in(StoreVectorScatterMaskedNode::Mask))) {
```
```suggestion
              mask->eqv_uncast(svgm->in(StoreVectorScatterMaskedNode::Mask))) {
```

eme64 commented:
Ah, some more missing cases:
- Do some store-store and store-load cases where you the first and second are different loads/stores, i.e. one with and one without mask/offsets. E.g. `StoreVectorMasked` and `StoreVectorScatter` in a store-store test. Doing the total cross-product is probably too much, but a few examples would be a good start.

dafedafe commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -3511,7 +3515,43 @@ Node* StoreNode::Identity(PhaseGVN* phase) {
+          result = mem;
+        }
+      }
+    }
```
I agree that it is quite convoluted probably also because I've put `if (!is_StoreVector())` (which is redundant) at the beginning to get the most common case out of the way but still...
At first I thought that multiple inheritance would be a good solution (masks and offsets could be inherited by the corresponding nodes) but the "HotSpot Coding Style" clearly says to avoid it...
So, I think in the end your second suggestion is the cleanest. Changing it...

dafedafe replied:
I've updated it. The condition unfortunately doesn't look as clean as the one above as we need to check for `nullptr` (either both or none and `eqv_uncast`). I've tried to make it as concise as possible (we could have made `mask` and `offsets` return a _unique_ node instead, so as to avoid the `nullptr`, but I had the impression it would just make everything less clear).

eme64 replied:
I think this is good, certainly much better than what we had before! :blush:

dafedafe commented:
> * No mixed type test for load-store: Use MemorySegment `from/intoMmemorySegment`. Try something like store a int-vector, and load a float-vector.

It looks as if load/stores that use `from`/`intoMemorySegment` with different types apparently don’t create `LoadVector` nodes. It seems that `fromMemorySegment` tries to inline the `VectorSupport::load` intrinsic, but fails as the type of the vector and  the inferred type of the underlying memory segment differ: https://github.com/openjdk/jdk/blob/9b742109b196d79cbf712ffd3f64edd1d6497114/src/hotspot/share/opto/vectorIntrinsics.cpp#L1055-L1064

> * Mismatched vector length: store a vector of length 4, and load one of length 8.

I've added tests tests that store and load with different species (`SPECIES_64`).

> * Do some store-store and store-load cases where you the first and second are different loads/stores, i.e. one with and one without mask/offsets. E.g. `StoreVectorMasked` and `StoreVectorScatter` in a store-store test. Doing the total cross-product is probably too much, but a few examples would be a good start.

You're right, there were just very few of them. Added many more.

eme64 commented:
> It looks as if load/stores that use from/intoMemorySegment with different types apparently don’t create LoadVector nodes. It seems that fromMemorySegment tries to inline the VectorSupport::load intrinsic, but fails as the type of the vector and the inferred type of the underlying memory segment differ:

Ha, that seems to be a bit of an arbitrary (maybe just conservative) restriction. I hope we can lift that in the future. We do not have such restrictions for scalar array load/store. And we can also Auto-Vectorize those. Would be interesting to ask some VectorAPI folks about that.

Looking at the tests now!

eme64 commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -974,6 +976,7 @@ class StoreVectorNode : public StoreNode {
    virtual uint match_edge(uint idx) const { return idx == MemNode::Address ||
                                                     idx == MemNode::ValueIn ||
                                                     idx == MemNode::ValueIn + 1; }
+   virtual Node* offsets() const { return in(Offsets); }
```
Would be nice to add some `override` keywords here ;)

eme64 replied:
And also below.

dafedafe replied:
On macosx we build with the flag `-Winconsistent-missing-override` enabled, which seems to allow either no `override` anywhere or in all places where a method actually overrides another one. So, in this case it seems that we would either need to add `override`s to `virtual int Opcode()`, `virtual uint match_edge(uint idx)`, `virtual Node* Ideal(PhaseGVN* phase, bool can_reshape)` etc. in all classes, or remove the 4 `override` keywords from the `indices` and `mask` methods. I might opt for the second choice. @eme64 what do you think?

TobiHartmann replied:
The second choice makes sense to limit the scope of the changes in this PR. If we still want to add the `override` keywords, we should do that in a separate RFE.

eme64 replied:
Ah yes. I forgot about `-Winconsistent-missing-override`. No `override` keyword then :disappointed:

dafedafe replied:
C'est la vie 😉 
Removed.
Thanks guys.

eme64 commented:
Nice, the tests look good, thanks for adding them!

dafedafe commented:
> Because we could allow those to use vectors in the future: I would leave the type checks in for now.

👍 better safe than sorry

eme64 reviewed (approved):
Looks good now! Thanks for all the updates, I think now the fix looks really concise :blush: 

> we add a specific store_Opcode method to LoadVectorGatherNode, LoadVectorMaskedNode and LoadVectorGatherMaskedNode that doesn’t return a store opcode but instead returns its own (to avoid ever being the same as a store node).

This part in the PR description could be updated: now we return `-1` for those that we think are not "comparable".

dafedafe commented:
> This part in the PR description could be updated: now we return `-1` for those that we think are not "comparable".

You're right. Fixed.
Thanks a lot for the review @eme64!!

TobiHartmann commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -890,6 +890,7 @@ class LoadVectorNode : public LoadNode {
 // Load Vector from memory via index map
 class LoadVectorGatherNode : public LoadVectorNode {
  public:
+  enum { Offsets = 3 };
```
Just wondering about the naming convention here because the corresponding constructor argument is called `indices`. Should that be consistent?

TobiHartmann replied:
Also, where is this used?

dafedafe replied:
You're right, they are always called "indices" even in the Vector API documentation. I've renamed them: offsets -> indices.
I've also removed all unused enums (in `LoadVector...Node` classes).

TobiHartmann commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -910,6 +911,10 @@ class LoadVectorGatherNode : public LoadVectorNode {
   }
+  virtual int store_Opcode() const {
+    // Ensure it is different from any store opcode
+    return -1;
```
Maybe improve this comment and explain that we are doing this to avoid folding which does not account for the mask/offsets. Same for the comments in the other `store_Opcode` methods.

dafedafe replied:
Comments updated.
Thanks @TobiHartmann!

TobiHartmann commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -1007,12 +1019,19 @@ class LoadVectorMaskedNode : public LoadVectorNode {
  public:
+  enum { Offsets = 3,
+         Mask
+  };
```
Where is this used?

dafedafe replied:
Removed.

TobiHartmann commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -987,12 +997,14 @@ class StoreVectorMaskedNode : public StoreVectorNode {
 // Load Vector from memory under the influence of a predicate register(mask).
 class LoadVectorMaskedNode : public LoadVectorNode {
  public:
+  enum { Mask = 3 };
```
Where is this used?

dafedafe replied:
Removed.

TobiHartmann reviewed (changes requested):
Nice regression test! I added a few comments.

eme64 commented:
@dafedafe I recommend that you update the Bug title. It names "offsets", but really you probably want to name "indices and mask", or just "masked and gather/scatter vector operations". And it also is not just about "loads", but also "stores", correct?

TobiHartmann reviewed (approved):
That looks good to me, thanks for adjusting!

dafedafe commented:
> @dafedafe I recommend that you update the Bug title. It names "offsets", but really you probably want to name "indices and mask", or just "masked and gather/scatter vector operations". And it also is not just about "loads", but also "stores", correct?

Yep. I changed the title... and adapted the description as well (offsets -> indices).
Thanks @eme64

TobiHartmann reviewed (approved):
This is good to go.

dafedafe commented:
/integrate
